miniapp/miniapp_cholesky.cpp

```diff
@@ -0,0 +1,240 @@
+    if (0 == world.rank())
+      std::cout << "[" << run_index << "]" << std::endl;
+
+    setup_input_matrix(matrix);
+
+    // wait all setup tasks before starting benchmark
+    {
+      for (int local_tile_j = 0; local_tile_j < distribution.localNrTiles().cols(); ++local_tile_j)
+        for (int local_tile_i = 0; local_tile_i < distribution.localNrTiles().rows(); ++local_tile_i)
+          matrix(LocalTileIndex{local_tile_i, local_tile_j}).get();
+      MPI_Barrier(world);
+    }
+
+    common::Timer<> timeit;
+    cholesky(comm_grid, blas::Uplo::Lower, matrix);
+
+    // wait for last task and barrier for all ranks
+    {
+      GlobalTileIndex last_tile(matrix.nrTiles().rows() - 1, matrix.nrTiles().cols() - 1);
+      if (matrix.rankIndex() == distribution.rankGlobalTile(last_tile))
+        matrix(last_tile).get();
+
+      MPI_Barrier(world);
+    }
+    auto elapsed_time = timeit.elapsed();
+
+    double gigaflops;
+    {
+      double n = matrix.size().rows();
+      double add_mul = n * n * n / 6;
+      gigaflops = total_ops<T>(add_mul, add_mul) / elapsed_time / 1e9;
+    }
+
+    // print benchmark results
+    if (0 == world.rank())
+      std::cout << "[" << run_index << "]"
+                << " " << elapsed_time << "s"
+                << " " << gigaflops << "GFlop/s"
+                << " " << matrix.size() << " " << matrix.blockSize() << " " << comm_grid.size() << " "
+                << hpx::get_os_thread_count() << std::endl;
+
+    // (optional) run test
+    if (opts.do_check)
+      cholesky_check(matrix);
+  }
+
+  return hpx::finalize();
+}
+
+int main(int argc, char** argv) {
+  // Initialize MPI
+  int threading_required = MPI_THREAD_SERIALIZED;
+  int threading_provided;
+  MPI_Init_thread(&argc, &argv, threading_required, &threading_provided);
+
+  if (threading_provided != threading_required) {
+    std::fprintf(stderr, "Provided MPI threading model does not match the required one.\n");
+    MPI_Abort(MPI_COMM_WORLD, 1);
+  }
+
+  // options
+  using namespace hpx::program_options;
+  options_description desc_commandline("Usage: " HPX_APPLICATION_STRING " [options]");
+
+  // clang-format off
+  desc_commandline.add_options()
+    ("matrix-size",
+     value<int64_t>()->default_value(4096),
+     "Matrix size.")
+    ("block-size",
+     value<int64_t>()->default_value(256),
+     "Block cyclic distribution size.")
+    ("grid-rows",
+     value<int64_t>()->default_value(1),
+     "Number of row processes in the 2D communicator.")
+    ("grid-cols",
+     value<int64_t>()->default_value(1),
+     "Number of column processes in the 2D communicator.")
+    ("nruns",
+     value<int64_t>()->default_value(1),
+     "Number of runs to compute the cholesky")
+    ("check-result",
+     bool_switch()->default_value(false),
+     "Check the cholesky factorization (for each run)")
+  ;
+  // clang-format on
+
+  // Create the resource partitioner
+  hpx::resource::partitioner rp(desc_commandline, argc, argv);
```

Comment:
In HPX things on resource partitioner side are changing:

> on 1.4.0 you can only create the rp upfront and not pass it to init, and on master it can only be passed to init (just creating it does nothing)

>for the next release we will most likely change it to be so that you can't even create it yourself but you only pass resource partitioner parameters to init

Moreover, there is a problem with the management of the help message, that it does not stop the application to run. More on this at https://github.com/STEllAR-GROUP/hpx/issues/4317